include/toolchain/common.h

```diff
@@ -144,4 +144,48 @@
 #define BUILD_ASSERT_MSG(EXPR, MSG) BUILD_ASSERT(EXPR)
 #endif
 
+/*
+ * This is meant to be used in conjonction with __in_section() and similar
+ * where scattered structure instances are concatened together by the linker
+ * and walked by the code at run time just like a contiguous array of such
+ * structures.
+ *
+ * Assemblers and linkers may insert alignment padding by default whose
+ * size is larger than the natural alignment for those structures when
+ * gathering various section segments together, messing up the array walk.
+ * To prevent this, we need to provide an explicit alignment not to rely
+ * on the default that might just work by luck.
+ *
+ * Alignment statements in  linker scripts are not sufficient as
+ * the assembler may add padding by itself to each segment when switching
+ * between sections within the same file even if it merges many such segments
+ * into a single section in the end.
+ */
+#define __decl_align(type) __aligned(__alignof(type)) type
+
+/*
+ * Convenience helper combining __in_section() and __decl_align().
+ * The section name is the struct type prepended with an underscore.
+ * The subsection is "static" and the subsubsection is the variable name.
+ */
+#define declare_struct_in_section(struct_type, name) \
```

Comment:
(whoops, dropped this comment, sorry)

This I like very much!

Style: Generally, zephyr code will put the verb "declare" at the end
of the symbol name, use upper case for macros, and add a z_ or Z_ prefix to
all internal symbols.  So maybe Z_STRUCT_SECTION_DECLARE() might be more conventional.

Or if I can suggest: maybe calling the feature "ITERABLE_STATIC" better captures the magic at work.  The point isn't that we're using structs (this trick works for any fixed-size data type) or that they're going in a particular section, it's that the link magically makes them into a list for us at runtime no matter where they might be.